### PR TITLE
Makes the blood filter not suck so much

### PR DESCRIPTION
**Pull request by GoblinBackwards**
## About The Pull Request
Adds a whitelist to the blood filter, you can use it in hand to bring up a UI similar to the plumbing filters. When using the blood filter for surgery it will only filter out chems in the whitelist, unless the list is empty in which case it will filter all chems.

Makes the surgery loop until all filterable chems are removed, similar to tend wounds.

Holding a health analyzer in your off hand while using the blood filter will display a chem scan of the subject after each surgery step similar to tend wounds.

## Why It's Good For The Game
In its current state the blood filter is just awful to use. It does nothing unique, is flat out worse at its job that most purging chems, you have to click the patient after every use even though it usually takes a lot of steps to actually filter out all the chems from someone, and you can't even see how much more you have to filter without pausing the surgery to swap to a health analyzer to chem scan separately.

## Changelog

:cl:
add: Added a whitelist to blood filters, unless the list is empty they will only remove chems in this whitelist. You can edit the list by using the filter in your hand
qol: Blood filter surgery now automatically does a chem scan after each step if you're holding a health analyzer.
qol: Blood filter surgery now automatically loops if the patient has any filterable chems in them
/:cl:

### CODE REVIEW

**Review by stylemistake — APPROVED** [review]

File: code/modules/surgery/blood_filter.dm [hunk 1]
```diff
@@ -16,6 +16,26 @@
...
 	return ..()
 
+/datum/surgery_step/filter_blood/initiate(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery, try_to_fail = FALSE)
+	if(!..())
+		return
+	while(has_whitelisted_chems(target, tool))
+		if(!..())
+			break
+
+/datum/surgery_step/filter_blood/proc/has_whitelisted_chems(mob/living/carbon/target, obj/item/blood_filter/bf)
```
Comment (MrMelbert) [inline code/modules/surgery/blood_filter.dm:26]: Auto doc this proc

Also, better variable name than "`bf`".

File: code/modules/surgery/blood_filter.dm [hunk 2]
```diff
@@ -16,6 +16,26 @@
...
+	if(!target.reagents?.reagent_list.len)
```
Comment (MrMelbert) [inline code/modules/surgery/blood_filter.dm:27]:
```suggestion
	if(!length(target.reagents?.reagent_list))
```

File: code/modules/surgery/blood_filter.dm [hunk 3]
```diff
@@ -16,6 +16,26 @@
...
+		return FALSE
+
+	if(!bf.whitelist_ids.len)
```
Comment (MrMelbert) [inline code/modules/surgery/blood_filter.dm:30]:
```suggestion
	if(!length(bf.whitelist_ids))
```

File: code/modules/surgery/blood_filter.dm [hunk 4]
```diff
@@ -16,6 +16,26 @@
...
+		return TRUE
+
+	for(var/datum/reagent/chem in target.reagents.reagent_list)
```
Comment (MrMelbert) [inline code/modules/surgery/blood_filter.dm:33]:
```suggestion
	for(var/datum/reagent/chem as anything in target.reagents.reagent_list)
```

File: code/modules/surgery/blood_filter.dm [hunk 5]
```diff
@@ -30,12 +50,18 @@
 	display_pain(target, "You feel a throbbing pain in your chest!")
 
 /datum/surgery_step/filter_blood/success(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery, default_display_results = FALSE)
+	var/obj/item/blood_filter/bf = tool
```
Comment (MrMelbert) [inline code/modules/surgery/blood_filter.dm:53]: Better variable name.

File: code/modules/surgery/tools.dm [hunk 6]
```diff
@@ -453,3 +453,39 @@
 	attack_verb_simple = list("pump", "siphon")
 	tool_behaviour = TOOL_BLOODFILTER
 	toolspeed = 1
+	var/list/whitelist_ids = list()
+	var/list/whitelist_english = list()
```
Comment (MrMelbert) [inline code/modules/surgery/tools.dm:457]: These need auto-doc comments as well.

File: code/modules/surgery/blood_filter.dm [hunk 7]
```diff
@@ -30,12 +50,18 @@
...
 	if(target.reagents?.total_volume)
 		for(var/datum/reagent/chem as anything in target.reagents.reagent_list)
-			target.reagents.remove_reagent(chem.type, min(chem.volume * 0.22, 10))
+			if(!bf.whitelist_ids.len || bf.whitelist_ids.Find(chem.type))
```
Comment (MrMelbert) [inline code/modules/surgery/blood_filter.dm:56]:
```suggestion
			if(!length(bf.whitelist_ids) || chem.type in bf.whitelist_ids)
```

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Pull request merged**